Support cursor visibility codes

kj7rrv opened the pull request:
Support cursor visibility codes; closes #16

kj7rrv committed:
Add cursor visbility classes

kj7rrv commented:
@schneiderfelipe when I call Ansi.escape(), the first Escape object is `Escape('\x1b[?25l\x1b[?7l')`. It looks like this is two codes. Is that normal?

schneiderfelipe commented:
> @schneiderfelipe when I call Ansi.escape(), the first Escape object is `Escape('\x1b[?25l\x1b[?7l')`. It looks like this is two codes. Is that normal?

Probably a bug, likely here https://github.com/getcuia/stransi/blob/6997722fb946aa8ac732b54e3fd623f87706013a/src/stransi/ansi.py#L33-L42 or here https://github.com/getcuia/stransi/blob/6997722fb946aa8ac732b54e3fd623f87706013a/src/stransi/escape.py#L20-L22

I feel stransi doesn't work well with `\x1b[?...`. It would be nice to test this.

kj7rrv commented:
It looks like isescape should be fine; the problem is probably in escapes.

kj7rrv commented:
It seems like something about `\x1b[?25l` makes the regex not recognize the end of the code, so it detects `\x1b[?7l` as part of the same code.

kj7rrv commented:
The `?` is the problem, like you thought.

    >>> list(stransi.Ansi('\x1b[?25l\x1b[?7l').escapes())
    [Escape('\x1b[?25l\x1b[?7l')]
    >>> list(stransi.Ansi('\x1b[25l\x1b[?7l').escapes())
    [Escape('\x1b[25l'), Escape('\x1b[?7l')]

kj7rrv committed:
Fix regex

kj7rrv commented:
I fixed the regex in 139aaca.

kj7rrv commented:
I'm not sure what exactly the `?` means; do you know @schneiderfelipe? Now the exception it gives on the neofetch output is `ValueError: invalid literal for int() with base 10: '?25'`.

kj7rrv commented:
Okay, it looks like `?` means it's a private sequence. To continue using ints, maybe we should represent codes with `?` as negative ints? e.g. `?25` becomes `-25`

kj7rrv committed:
Use negative integers for private values

Private values (starting with `?`) are represented as negative integers;
for example, `?25` becomes -25.

kj7rrv commented:
@schneiderfelipe ab339f9 uses negative ints for private sequences, so now the private codes are properly handled as unsupported codes.

kj7rrv committed:
Add support for cursor visibility changes

kj7rrv commented:
Okay, I think this is ready to merge.

schneiderfelipe commented on src/stransi/escape.py:
```diff
@@ -39,7 +39,7 @@ def tokens(self) -> Iterator[Token]:
                 yield Token(kind=kind, data=0)
                 continue
-            yield Token(kind=kind, data=int(param))
+            yield Token(kind=kind, data=int(param.replace('?', '-')))
```
I don't like that. It would be better if we had a boolean flag in `Token` https://github.com/getcuia/stransi/blob/c4a395fbe9144574cc8fd48b3eb99a502de47395/src/stransi/token.py#L9-L40

schneiderfelipe replied:
`is_private` with a false default value, true for private codes. What do you think? @kj7rrv

kj7rrv replied:
The only problem with that is an existing program looking for a non-private code would start detecting the private code with the same number.

schneiderfelipe replied:
Why is that? Couldn't they check the `is_private` flag?

My concern is that we might need to add yet another dimension to `Token` in the future, so mapping to different ranges of integers doesn't scale well.

kj7rrv replied:
They could, but existing programs don't. Maybe both? Negative means "This isn't a normal non-private code", and then they have to check why?

schneiderfelipe replied:
Which existing programs?

kj7rrv replied:
I don't know of any particular ones, but this could affect any program that uses stransi, checks the `data` value to determine what to do with a code ,and sometimes receives private codes as input. I can do this if you want, but it will be an API-breaking change.

schneiderfelipe replied:
That's OK, we're under 1.0.0 at the moment.

kj7rrv replied:
Okay, I'll do it. It looks like not all private codes use a `?` (most do, but a few use other characters). I'll leave the detection code as-is for now, but I'll add a `private_designator` attribute to `Token` that will always be `?` for now (if the token is private; it will be empty for non-private tokens), in addition to `is_private`.

kj7rrv replied:
Done in 5efbe17

schneiderfelipe commented on src/stransi/cursor.py:
```diff
@@ -58,3 +58,27 @@ class SetCursor(Instruction[CursorMove]):
+
+
+@dataclass
+class CursorVisibilityChange:
```
Could you change the name of this class to `CursorVisibility`?

kj7rrv replied:
Done in 2586d3d

schneiderfelipe reviewed (changes requested):
Looks good, just a simple request. 👍🏽

kj7rrv committed:
CursorVisibilityChange -> CursorVisibility

kj7rrv commented:
I don't know of any particular ones, but this could affect any program that uses stransi, checks the `data` value to determine what to do with a code ,and sometimes receives private codes as input. I can do this if you want, but it will be an API-breaking change.

kj7rrv committed:
is_private and private_designator
